### ISSUE TITLE
dependency-check-maven:6.0.3 complains about vulnerable dependencies

### ISSUE THREAD

**Issue by AxelNennker** [vaadin/framework#12176]
dependency-check-maven:6.0.3 complains about vulnerable dependencies.

```
[ERROR] Failed to execute goal org.owasp:dependency-check-maven:6.0.3:check (default-cli) on project vaadin-push: 
[ERROR] 
[ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '8.0': 
[ERROR] 
[ERROR] vaadin-slf4j-jdk14-1.6.1.jar: CVE-2018-8088
```

How to reproduce.
- clone the framework
- use jdk8 because framework does not build with jdk15
- insert dependency-check-maven plugin into pom.xml
```
                <plugin>
                  <groupId>org.owasp</groupId>
                  <artifactId>dependency-check-maven</artifactId>
                  <version>6.0.3</version>
                    <configuration>
                        <failBuildOnCVSS>8</failBuildOnCVSS>
                    </configuration>
                  <executions>
                      <execution>
                          <goals>
                              <goal>check</goal>
                          </goals>
                      </execution>
                  </executions>
                </plugin>
```
- run `mvn clean install`
- run `mvn clean org.owasp:dependency-check-maven:6.0.3:check`

This outputs
```
[ERROR] One or more dependencies were identified with vulnerabilities that have a CVSS score greater than or equal to '8.0': 
[ERROR] 
[ERROR] vaadin-slf4j-jdk14-1.6.1.jar: CVE-2018-8088
```

`mvn versions:display-dependency-updates ` shows that many dependencies can be updates. Also some plugins can be updated.

I don't know whether updating the dependencies helps with resolving the issues dependency-check-maven complains about.
There might be false positives.

```
[INFO]   com.carrotsearch:smartsprites ....................... 0.2.10 -> 0.2.13
[INFO]   com.github.siom79.japicmp:japicmp-maven-plugin ....... 0.9.1 -> 0.14.4
[INFO]   com.google.appengine:appengine-api-1.0-sdk ........... 1.7.7 -> 1.9.84
[INFO]   com.google.gwt:gwt-dev ................................ 2.8.2 -> 2.9.0
[INFO]   com.google.gwt:gwt-elemental .......................... 2.8.2 -> 2.9.0
[INFO]   com.google.gwt:gwt-servlet ............................ 2.8.2 -> 2.9.0
[INFO]   com.vaadin:vaadin-testbench .......................... 5.2.0 -> 18.0.3
[INFO]   commons-cli:commons-cli ................................... 1.2 -> 1.4
[INFO]   commons-codec:commons-codec .............................. 1.5 -> 1.15
[INFO]   commons-io:commons-io ................................... 2.4 -> 2.8.0
[INFO]   javax.portlet:portlet-api ............................... 2.0 -> 3.0.1
[INFO]   javax.servlet:javax.servlet-api ....................... 3.0.1 -> 4.0.1
[INFO]   org.easymock:easymock ..................................... 3.0 -> 4.2
[INFO]   org.eclipse.jetty:jetty-proxy ............. 9.4.35.v20201120 -> 11.0.0
[INFO]   org.eclipse.jetty:jetty-server ............ 9.4.35.v20201120 -> 11.0.0
[INFO]   org.eclipse.jetty:jetty-servlets .......... 9.4.35.v20201120 -> 11.0.0
[INFO]   org.eclipse.jetty:jetty-util .............. 9.4.35.v20201120 -> 11.0.0
[INFO]   org.eclipse.jetty:jetty-webapp ............ 9.4.35.v20201120 -> 11.0.0
[INFO]   org.eclipse.jgit:org.eclipse.jgit ...
[INFO]                            3.5.1.201410131835-r -> 5.10.0.202012080955-r
[INFO]   org.hibernate:hibernate-validator ........... 5.3.6.Final -> 7.0.0.CR1
[INFO]   org.hsqldb:hsqldb ..................................... 2.2.6 -> 2.5.1
[INFO]   org.jsoup:jsoup ..................................... 1.11.2 -> 1.13.1
[INFO]   org.mockito:mockito-core ............................. 1.9.5 -> 3.6.28
[INFO]   org.osgi:osgi.annotation .............................. 6.0.1 -> 8.0.0
[INFO]   org.osgi:osgi.cmpn .................................... 6.0.0 -> 7.0.0
[INFO]   org.osgi:osgi.core .................................... 6.0.0 -> 8.0.0
[INFO]   org.slf4j:slf4j-log4j12 ....................... 1.7.30 -> 2.0.0-alpha1
```

[Sormas](https://github.com/hzi-braunschweig/SORMAS-Project) is introducing dependency-check-maven and using vaadin8. It seems that vulnerabilities found there stem from dependencies of vaadin8.

ps: building seems to reformat a lot of files.

**Comment by Legioth** [comment]
The vulnerability report is a false positive in this case. 

Vaadin 8 uses `java.util.logging` as its native logging library (newer versions of Vaadin use SLF4J instead). At the same time, Vaadin uses Atmosphere for server push and Atmosphere is logging through SLF4J. To make logging from Atmosphere consistent with logging from the rest of Vaadin, there's a custom `vaadin-slf4j-jdk14` dependency that combines the SLF4J API and the adapter that forwards messages to `java.util.logging`. Furthermore, the Atmosphere version used by Vaadin is modified to use this custom logging API instead of the regular SLF4J API.

The CVE-2018-8088 vulnerability is related to a class in the `slf4j-ext` module. This module is not even included in `vaadin-slf4j-jdk14` which means that `vaadin-slf4j-jdk14` cannot be vulnerable to CVE-2018-8088. The OWASP scanner doesn't make that distinction but instead reacts to anything related to older versions of SLF4J.

I will leave this issue open for now not because there would be a known vulnerability that would need fixing, but rather to track work towards making changes that would prevent this situation from being misidentified.

**Comment by TatuLund** [comment]
This false positive report is known already for a while. We use OWASP tool (and other similar tools) periodically to check the current output. We have also bot checking dependencies directly from our GitHub repositories.

The code changes to update relevant SLF4J dependencies have been made already a while ago, but we have not yet released and new version of our Atmosphere fork

https://github.com/vaadin/atmosphere/commit/378e5f9a975b7a6cfb00f6f6aefbe39a9f94eb10

**Comment by AxelNennker** [comment]
Would you consider providing a suppression-file for vaadin8 to be used with dependency-check-maven?
```
<suppressionFiles>
  <suppressionFile>project-suppression.xml</suppressionFile>
</suppressionFiles>
```
The dependency-check report has an option `generate exclude-file` which makes this quite easy.

Anyways, updating all dependencies from time to time is a good thing, I think.
At least the minor version updates...

Thanks for looking into this.

**Comment by TatuLund** [comment]
> Would you consider providing a suppression-file for vaadin8 to be used with dependency-check-maven?

I would suggest to configure the OWASP plugin in your actual target application and use `<skipProvidedScope>true</skipProvidedScope>` option in order to avoid further false positives. See example here: https://github.com/TatuLund/cdi-demo/blob/master/pom.xml#L172

**Comment by TatuLund** [comment]
When I am running the correctly configured tool against Vaadin 8.12.1 based test app, I get the following result.

```
Dependency | Vulnerability IDs | Package | Highest Severity | CVE Count | Confidence | Evidence Count
-- | -- | -- | -- | -- | -- | --
deltaspike-core-api-1.7.2.jar | cpe:2.3:a:apache:deltaspike:1.7.2:*:*:*:*:*:*:* | pkg:maven/org.apache.deltaspike.core/deltaspike-core-api@1.7.2 | MEDIUM | 1 | Highest | 41
vaadin-cdi-3.0.1.jar | cpe:2.3:a:vaadin:vaadin:3.0.1:*:*:*:*:*:*:* | pkg:maven/com.vaadin/vaadin-cdi@3.0.1 | MEDIUM | 1 | Highest | 31
vaadin-slf4j-jdk14-1.6.1.jar | cpe:2.3:a:slf4j:slf4j-ext:1.6.1:*:*:*:*:*:*:* | pkg:maven/com.vaadin.external.slf4j/vaadin-slf4j-jdk14@1.6.1 | CRITICAL | 1 | Highest | 29
```

All three are false positives.

**Comment by AxelNennker** [comment]
Thanks for the advice with `<skipProvidedScope>true</skipProvidedScope>` Something for [Sormas](https://github.com/hzi-braunschweig/SORMAS-Project) to look into.
Although I think that updating dependency versions is beneficial most of the times, like 10^12 to 1 or something.
Maybe updating vaadin-slf4j*-1.6.1* to vaadin-slf4j*-1.7.30* makes the false positive go away.
https://mvnrepository.com/artifact/org.slf4j/slf4j-ext slf4j should be 1.7.30 in all vaadin repositories. The same for htmlunit and gwt.

Additionally, attacks on the build system have got more attention through the [Solarwinds](https://www.schneier.com/blog/archives/2020/12/how-the-solarwinds-hackers-bypassed-duo-multi-factor-authentication.html) incident. So, I think, saying that a vulnerability is compile/build-time only and therefor harmless neglects that now the developers have to be careful which PR they try out on their machine because that PR could include input that triggers such build-time vulnerability and then the developers machine or the CI system are compromised.

So updating has, in my opinion, all the pros and nearly never any cons.

**Comment by TatuLund** [comment]
> Maybe updating vaadin-slf4j*-1.6.1* to vaadin-slf4j*-1.7.30*

This is already work in progress (or actually to 1.8.beta4), see my earlier comment https://github.com/vaadin/framework/issues/12176#issuecomment-750146800

**Comment by AxelNennker** [comment]
Hi, any news on when the work in progress might land?
Also com.google.gwt could be version 2.9.0 in framework
```
<!-- https://mvnrepository.com/artifact/com.google.gwt/gwt-dev -->
<dependency>
    <groupId>com.google.gwt</groupId>
    <artifactId>gwt-dev</artifactId>
    <version>2.9.0</version>
    <scope>provided</scope>
</dependency>
```
Thank you for your support

**Comment by Ansku** [comment]
Tatu knows better about the status of that Atmosphere fix, but as for GWT 2.9, it drops some things that we use internally, so we can't upgrade the framework itself to it. Last I tried it seemed fairly straightforward to upgrade it locally, though -- you can replace the missing CheckForUpdates class with an empty implementation, and exclude the apache-jsp dependency that comes with the gwt-dev.

**Comment by TatuLund** [comment]
I just want to point out that using GWT 2.8.2 does not include vulnerability risk. So it is purely a technical update from our perspective, which is not at this point of urgent or mandatory from our perspective. GWT 2.9 mainly gives compatibility to newer Java versions and we have been studying it from that perspective and may have activities related to that later, once the internal issues referred by Anna are no longer concern for us. However we will in near future post instructions how to override your project to use it instead of framework provided version.
